File: server/cluster.go
package server

import (
	"errors"
	"fmt"
	"time"

	"github.com/RocksLabs/kvrocks_controller/controller/failover"
	"github.com/RocksLabs/kvrocks_controller/util"
	"github.com/gin-gonic/gin"

	"github.com/RocksLabs/kvrocks_controller/consts"
	"github.com/RocksLabs/kvrocks_controller/metadata"
	"github.com/RocksLabs/kvrocks_controller/storage"
)

type CreateClusterRequest struct {
	Name     string   `json:"name"`
	Nodes    []string `json:"nodes"`
	Password string   `json:"password"`
	Replicas int      `json:"replicas"`
}

func (req *CreateClusterRequest) validate() error {
	if len(req.Name) == 0 {
		return fmt.Errorf("cluster name should NOT be empty")
	}
	if len(req.Nodes) == 0 {
		return errors.New("cluster nodes should NOT be empty")
	}
	if !util.IsUniqueSlice(req.Nodes) {
		return errors.New("cluster nodes should NOT be duplicated")
	}

	invalidNodes := make([]string, 0)
	for _, node := range req.Nodes {
		if !util.IsHostPort(node) {
			invalidNodes = append(invalidNodes, node)
		}
	}
	if len(invalidNodes) > 0 {
		return fmt.Errorf("invalid node addresses: %v", invalidNodes)
	}

	if req.Replicas == 0 {
		req.Replicas = 1
	}
	if len(req.Nodes)%req.Replicas != 0 {
		return errors.New("cluster nodes should be divisible by replica")
	}
	return nil
}

type ClusterHandler struct {
	storage *storage.Storage
}

func (handler *ClusterHandler) List(c *gin.Context) {
	namespace := c.Param("namespace")
	clusters, err := handler.storage.ListCluster(c, namespace)
	if err != nil {
		responseError(c, err)
		return
	}
	responseOK(c, gin.H{"clusters": clusters})
}

func (handler *ClusterHandler) Get(c *gin.Context) {
	namespace := c.Param("namespace")
	clusterName := c.Param("cluster")
	cluster, err := handler.storage.GetClusterInfo(c, namespace, clusterName)
	if err != nil {
		responseError(c, metadata.ErrEntryNoExists)
		return
	}
	responseOK(c, gin.H{"cluster": cluster})
}

func (handler *ClusterHandler) Create(c *gin.Context) {
	namespace := c.Param("namespace")

	var req CreateClusterRequest
	if err := c.BindJSON(&req); err != nil {
		responseBadRequest(c, err)
		return
	}
	if err := req.validate(); err != nil {
		responseBadRequest(c, err)
		return
	}

	if c.GetHeader(consts.HeaderDontDetectHost) != "true" {
		for _, node := range req.Nodes {
			if err := util.DetectClusterNode(c, &metadata.NodeInfo{
				Addr:     node,
				Password: req.Password,
			}); err != nil {
				responseBadRequest(c, err)
				return
			}
		}
	}

	replicas := req.Replicas
	shards := make([]metadata.Shard, len(req.Nodes)/replicas)
	slotRanges := metadata.SpiltSlotRange(len(shards))
	for i := range shards {
		shards[i].Nodes = make([]metadata.NodeInfo, 0)
		for j := 0; j < replicas; j++ {
			nodeAddr := req.Nodes[i*replicas+j]
			role := metadata.RoleMaster
			if j != 0 {
				role = metadata.RoleSlave
			}
			shards[i].Nodes = append(shards[i].Nodes, metadata.NodeInfo{
				ID:        util.GenerateNodeID(),
				Addr:      nodeAddr,
				Password:  req.Password,
				Role:      role,
				CreatedAt: time.Now().Unix(),
			})
		}
		shards[i].SlotRanges = append(shards[i].SlotRanges, slotRanges[i])
		shards[i].MigratingSlot = -1
		shards[i].ImportSlot = -1
	}
	err := handler.storage.CreateCluster(c, namespace, &metadata.Cluster{
		Version: 1,
		Name:    req.Name,
		Shards:  shards,
	})
	if err != nil {
		responseError(c, err)
		return
	}
	responseCreated(c, "created")
}

func (handler *ClusterHandler) Remove(c *gin.Context) {
	namespace := c.Param("namespace")
	cluster := c.Param("cluster")
	err := handler.storage.RemoveCluster(c, namespace, cluster)
	if err != nil {
		responseError(c, err)
		return
	}
	responseOK(c, "ok")
}

func (handler *ClusterHandler) GetFailOverTasks(c *gin.Context) {
	namespace := c.Param("namespace")
	cluster := c.Param("cluster")
	typ := c.Param("type")
	failover, _ := c.MustGet(consts.ContextKeyFailover).(*failover.FailOver)
	tasks, err := failover.GetTasks(c, namespace, cluster, typ)
	if err != nil {
		responseError(c, err)
		return
	}
	responseOK(c, gin.H{"tasks": tasks})
}

File: web/src/entitits/Cluster.ts
export class Cluster {
    name: string;
    namespace: string;
    nodes: Node[];
    password: string|null;
    replicas: number|null;
    constructor(data?: object) {
        this.name = '';
        this.namespace = '';
        this.nodes = [];
        this.password = null;
        this.replicas = null;
        if(data) {
            if('name' in data && typeof data.name == 'string') {
                this.name = data.name;
            }
            if('namespace' in data && typeof data.namespace == 'string') {
                this.namespace = data.namespace;
            }
            if('nodes' in data && Array.isArray(data.nodes)) {
                this.nodes = data.nodes.filter(n => typeof n == 'string').map(n => new Node(n));
            }
            if('password' in data && typeof data.password == 'string') {
                this.password = data.password;
            }
            if('replicas' in data && typeof data.replicas == 'number') {
                this.replicas = Math.round(data.replicas);
            }
        }
    }
    getCreationBody(): object {
        const result: any = {
            name: this.name,
            nodes: this.nodes.map(n => n.addr)
        };
        if(this.password != null) {
            result.password = this.password;
        }
        if(this.replicas != null) {
            result.replicas = this.replicas;
        }
        return result;
    }
    parseFromResponse(obj: any) {
        if(typeof obj != 'object') {
            return;
        }
        if(typeof obj.name == 'string') {
            this.name = obj.name;
        }
        if(Array.isArray(obj.shards)) {
            this.nodes = obj.shards.map((item: any) => {
                const node = new Node();
                node.parseFromResponse(item);
                return node;
            });
        }
    }
}

class Node {
    id: string;
    addr: string;
    role: string;
    password: string;
    slotRanges: string[];
    constructor(addr?: string) {
        this.id = '';
        this.addr = '';
        this.role = '';
        this.password = '';
        this.slotRanges = [];
        if(typeof addr == 'string') {
            this.addr = addr;
        }
    }
    parseFromResponse(obj: any) {
        if(typeof obj != 'object') {
            return;
        }
        if(Array.isArray(obj.nodes) && obj.nodes.length > 0 && typeof obj.nodes[0] == 'object') {
            (typeof obj.nodes[0].id == 'string') && (this.id = obj.nodes[0].id);
            (typeof obj.nodes[0].addr == 'string') && (this.addr = obj.nodes[0].addr);
            (typeof obj.nodes[0].role == 'string') && (this.role = obj.nodes[0].role);
            (typeof obj.nodes[0].password == 'string') && (this.password = obj.nodes[0].password);
        }
        if(Array.isArray(obj.slot_ranges) && obj.slot_ranges.length > 0) {
            this.slotRanges = obj.slot_ranges.filter((item: any) => typeof item == 'string');
        }
    }
}